net/lnet/linux_kernel/klnet_drv.c

```diff
@@ -1164,20 +1164,20 @@ static int nlx_dev_ioctl_nidstrs_get(struct nlx_kcore_domain *kd,
 	for (i = 0, sz = 1; nidstrs[i] != NULL; ++i)
```

Comment:
Here the check:
```
char **nidstrs;
...
nidstrs[i] != NULL;
```

`nidstrs` is an array of pointer to "char *", which is allocated in nlx_core_nidstrs_get().
But as I see, the array has the exact number of item with the nid.
This `nidstrs` has every item pointing to a string, and even the last item is pointing to a string.
If we check "nidstrs[i] != NULL", it will go beyond the last item. (out of array boundary)
This is because we don't have a size of this array.

Two ways to fix this:
- add a size/len for this array.
- place a NULL at the end of this array as an anchor. One more item should be allocated in nlx_core_nidstrs_get() and NULL assigned to it.

Reply:
> But as I see, the array has the exact number of item with the nid.
```
1753         for (nr = 0, rc = 0; rc != -ENOENT; ++nr)
1754                 rc = LNetGetId(nr, &id);
```
Let's assume there is 1 LNID in the list.
First iteration: nr = 0, rc = 0, rc != -ENOENT, rc = 0, nr = 1
Second iteration: rc != -ENOENT, rc = -ENOENT, nr = 2
Loop breaks.
So the amount of entries allocated would be 2. All zeroed by `m0_alloc()`. Only one entry will be filled, so the last one will be NULL. So your second suggestion is implemented already.

Reply:
Yes, you are right.
Can you consider adding this:

```
for (i = 0; i < nr - 1; ++i) {
		rc = LNetGetId(i, &id);
		M0_ASSERT(rc == 0);
		nidstr = libcfs_nid2str(id.nid);
		M0_ASSERT(nidstr != NULL);
		(*nidary)[i] = m0_strdup(nidstr);
		if ((*nidary)[i] == NULL) {
			for (j = 0; j < i; ++j)
				m0_free((*nidary)[j]);
			return M0_ERR(-ENOMEM);
		}
	}
(*nidary)[i] = NULL; /* the last item of this array needs NULL */
```

I know this is not needed. But it helps to understand this piece of code better.
Or this can be a comment.

Reply:
We can save on `j` variable here like this:

```
		if ((*nidary)[i] == NULL) {
			while (--i >= 0)
				m0_free((*nidary)[i]);
			return M0_ERR(-ENOMEM);
		}
```